File: array/18/main.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

// Function declaration
int** fourSum(int* nums, int numsSize, int target, int* returnSize, int** returnColumnSizes);

// Helper function to print array
void printArray(int* arr, int size) {
    printf("[");
    for (int i = 0; i < size; i++) {
        printf("%d", arr[i]);
        if (i < size - 1) printf(", ");
    }
    printf("]");
}

// Helper function to print result
void printResult(int** result, int returnSize, int* returnColumnSizes) {
    printf("[");
    for (int i = 0; i < returnSize; i++) {
        printArray(result[i], returnColumnSizes[i]);
        if (i < returnSize - 1) printf(", ");
    }
    printf("]\n");
}

void runTest(int* nums, int numsSize, int target, int expectedSize, int testCase) {
    printf("Test case %d:\n", testCase);
    printf("nums = ");
    printArray(nums, numsSize);
    printf(", target = %d\n", target);
    
    int returnSize;
    int* returnColumnSizes;
    int** result = fourSum(nums, numsSize, target, &returnSize, &returnColumnSizes);
    
    printf("Result: ");
    printResult(result, returnSize, returnColumnSizes);
    printf("Return size: %d\n", returnSize);
    
    // Assert the return size matches expected size
    assert(returnSize == expectedSize);
    
    // Free allocated memory
    for (int i = 0; i < returnSize; i++) {
        free(result[i]);
    }
    free(result);
    free(returnColumnSizes);
    
    printf("Test case %d passed!\n\n", testCase);
}

int main() {
    // Test case 1: Example from the problem
    int nums1[] = {1, 0, -1, 0, -2, 2};
    runTest(nums1, 6, 0, 3, 1);  // Expected 3 quadruplets
    
    // Test case 2: All same numbers
    int nums2[] = {2, 2, 2, 2, 2};
    runTest(nums2, 5, 8, 1, 2);  // Expected 1 quadruplet
    
    // Test case 3: No solution
    int nums3[] = {1, 2, 3, 4};
    runTest(nums3, 4, 100, 0, 3);  // Expected 0 quadruplets
    
    printf("All test cases passed!\n");
    return 0;
}

File: array/18/Main.java
import java.util.*;

public class Main {
    public static void testFourSum() {
        Solution solution = new Solution();
        
        // Test case 1: Example from the problem
        int[] nums1 = {1, 0, -1, 0, -2, 2};
        int target1 = 0;
        List<List<Integer>> expected1 = Arrays.asList(
            Arrays.asList(-2, -1, 1, 2),
            Arrays.asList(-2, 0, 0, 2),
            Arrays.asList(-1, 0, 0, 1)
        );
        List<List<Integer>> result1 = solution.fourSum(nums1, target1);
        System.out.println("Test case 1:");
        System.out.println("nums=" + Arrays.toString(nums1));
        System.out.println("target=" + target1);
        System.out.println("expected=" + expected1);
        System.out.println("result=" + result1);
        assert result1.size() == expected1.size() : "Test case 1 failed: Wrong number of quadruplets";
        
        // Test case 2: All same numbers
        int[] nums2 = {2, 2, 2, 2, 2};
        int target2 = 8;
        List<List<Integer>> expected2 = Arrays.asList(
            Arrays.asList(2, 2, 2, 2)
        );
        List<List<Integer>> result2 = solution.fourSum(nums2, target2);
        System.out.println("\nTest case 2:");
        System.out.println("nums=" + Arrays.toString(nums2));
        System.out.println("target=" + target2);
        System.out.println("expected=" + expected2);
        System.out.println("result=" + result2);
        assert result2.size() == expected2.size() : "Test case 2 failed: Wrong number of quadruplets";
        
        System.out.println("\nAll test cases passed!");
    }
    
    public static void main(String[] args) {
        testFourSum();
    }
}